keps/sig-api-machinery/20190226-network-proxy.md

```diff
@@ -0,0 +1,618 @@
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+The plan is to build a system which allows us to control where the traffic
+coming out of the Kube API Server is sent in an extensible manner. We will do
+this by adding a network proxy system. The KAS can be configured to send traffic
+(or not) to one or more of the proxies. These proxies can then be used as
+extensibility points for that traffic. In addition they will allow for control
+of the network.
+
+## Motivation
+
+Everyone [hates the SSH tunnels](https://github.com/kubernetes/kubernetes/issues/54076). They complicate kube-apiserver code a fair amount and only one cloud provider ever bothered to implement them.
+After a generous year of deprecation time, they will be removed in an upcoming release.
+
+In retrospect, having an explicit level of indirection that separates user-initiated network traffic from API server-initiated traffic seems to be a useful concept.
+This gives Cloud providers the ability to control how API server to pod, node and service connectivity is implemented.
+For example, a Cloud provider may choose to run their API server (control network) and the cluster nodes (cluster network) on isolated private IP networks with overlapping
+IP addresses and require custom communications implementation to reach the endpoints.
+Such a layer would also enable audit logging of requests made by the API server and allows for more validation of outgoing API server connections.
```

Comment:
"Such a layer" -> "An indirection layer" as I don't think you've mentioned the layer yet